net.solarnetwork.node.setup.web/web/js/graph-data.js

```diff
@@ -0,0 +1,237 @@
+//Author Robert Rewcastle
+//script to draw line graphs of datums with a wattage variable
+//this script will put the graphs inside an html tag with a graphpoint class which 
+//one must put in the html document before loading the script
+//requires d3 https://d3js.org/d3.v3.js
+
+//json field we are interested in
+var key = "watts";
+var units = "W";
+
+//This map is graphs to look up the latest reading based on their sourceId
+var datamap = {};
+
+//this array will contain all the sourceIds of graphs we are plotting
+var sourceIds = [];
+
+//https://stackoverflow.com/questions/1960473/get-all-unique-values-in-an-array-remove-duplicates
+//This get used for deciding how many ticks on the y axis
+function onlyUnique(value, index, self) { 
+    return self.indexOf(value) === index;
+}
+
+//https://stackoverflow.com/questions/7343890/standard-deviation-javascript
+//This get used for giving the graph some white space either side of the graph
+function StandardDeviation(numbersArr) {
+    //--CALCULATE AVAREGE--
+    var total = 0;
+    for(var key in numbersArr) 
+       total += numbersArr[key];
+    var meanVal = total / numbersArr.length;
+    //--CALCULATE AVAREGE--
+  
+    //--CALCULATE STANDARD DEVIATION--
+    var SDprep = 0;
+    for(var key in numbersArr) 
+       SDprep += Math.pow((parseFloat(numbersArr[key]) - meanVal),2);
+    var SDresult = Math.sqrt(SDprep/numbersArr.length);
+    //--CALCULATE STANDARD DEVIATION--
+    return SDresult
+    
+}
+
+//when a new datum comes in, this handler gets called
+//the handler checks if the new datum has a wattage reading and 
+var handler = function handleMessage(msg) {
+	
+	var datum = JSON.parse(msg.body).data;
+
+	//check that the datum has a reading
+	if (datum[key] != undefined){
+		//map the sourceId to the current reading
+		datamap[datum.sourceId] = datum[key];
+		
+		//if we have not seen this sourceId before we need to graph it
+		if (sourceIds.indexOf(datum.sourceId)==-1){
+			
+			//add a new graph for this new sourceId
+			sourceIds.push(datum.sourceId);
+			graphinit(datum.sourceId,units);
+		}
+	}
+};
+
+//creates a new graph looking that plots wattage data from datums coming from source 
+//this code is heavily based on the last graph from https://bost.ocks.org/mike/path/
+//apologies for the magic numbers 
+function graphinit(source, units) {
+
+
+    var n = 243,//how many sample points to have on the graph
+        duration = 1000,//time for the animation 
+        now = new Date(Date.now() - duration),//not sure what the -duration is for
+
+        //prefill the array with the first reading
+        data = new Array(n).fill(datamap[source]);
+
+    //positional styling for the graph
+    var margin = { top: 10, right: 0, bottom: 20, left: 60 },
+        width = 960 - margin.right,
```

Comment:
The width is too wide here. The charts should be able to render without horizontal scrolling on small devices like tablets. Given datum values often only chance once/minute a much narrower chart would be fine here anyway. Something like `320` might be plenty.